### PR TITLE
(router): allow pass array of routes to routes config

### LINKED ISSUE #9876: Router:  provideRouter(config) should spread route definition arrays for us
**I'm submitting a ...**  (check one with "x")

```
[ ] bug report
[x] feature request
[ ] support request 
```

**Current behavior**

When accumulating routes today we write this:

```
export const routes: RouterConfig = [
  ...HeroesRoutes,
  ...LoginRoutes,
  ...CrisisCenterRoutes
];

export const APP_ROUTER_PROVIDERS = [  provideRouter(routes), ...];
```

It's easy for developers to forget the spread operator `...`. They shouldn't have to worry about it just as they don't worry about it when setting up the `providers` array for component metadata.

**Expected/desired behavior**

We'd like to write them like this:

```
export const routes: RouterConfig = [
  HeroesRoutes,
  LoginRoutes,
  CrisisCenterRoutes
];
```

... and rely on `provideRouter` to spread them out.

**Reproduction of the problem**

See `app.routes.ts` in [plunker for router chapter](https://angular.io/resources/live-examples/router/ts/plnkr.html)

### PR DESCRIPTION
**Pull request by reneweb**
**Please check if the PR fulfills these requirements**
- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit-message-format
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

**What kind of change does this PR introduce?** (check one with "x")

```
[ ] Bugfix
[x] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[ ] Other... Please describe:
```

**What is the current behavior?** (You can also link to an open issue here)
#9876

**What is the new behavior?**
Routes is now of type Array<Route|Route[]> and flattened into Route[] in the router.

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[x] No
```

If this PR contains a breaking change, please describe the impact and migration path for existing applications: ...

**Other information**:

Closes #9876

### CODE REVIEW

File: modules/@angular/router/src/router.ts [hunk 1]
```diff
@@ -297,6 +298,18 @@ export class Router {
...
+  private flattenRouteConfig(routes: Routes): Route[] {
+    return routes
+        .map(function(e) {
+          if (e instanceof Array) {
+            return e;
+          } else {
+            return [e];
+          }
+        })
+        .reduce(function(prev, curr) { return prev.concat(curr); });
```
Comment (sod) [inline modules/@angular/router/src/router.ts]: Well, or just do `[].concat(...routes)`

File: modules/@angular/router/src/router.ts [hunk 2]
```diff
@@ -193,8 +193,9 @@ export class Router {
    * ```
    */
   resetConfig(config: Routes): void {
-    validateConfig(config);
-    this.config = config;
+    let flattenedConfig = this.flattenRouteConfig(config);
```
Comment (sod) [inline modules/@angular/router/src/router.ts]: Or just do `let flattenedConfig = [].concat(...config);`

Comment (reneweb) [inline modules/@angular/router/src/router.ts]: Yeah, makes it much simpler. I've pushed the change just now.

**Comment by vsavkin** [comment]
Sorry for a very late reply.

I think a better way to do it would be to update the Route interface to look like this:

```
type RouteChildren = Array<Route|RouteChildren>;
export interface Route {
  children?: RouteChildren;
}
```

And then update `applyRedirects` and `recognize` to flatten the tree into an array inside `getChildConfig`.

**Comment by reneweb** [comment]
No problem, almost forgot about the pr myself :)

I've started to look into your suggested approach, but  `type RouteChildren = Array<Route|RouteChildren>;` gives a circular reference error:
`config.ts:461:13: Type alias 'RouteChildren' circularly references itself.`
So, seems like we need to find something different.

**Comment by Toxicable** [comment]
Any updates on this? @reneweb are you planning on picking this back up again?

**Comment by reneweb** [comment]
I am not sure if this pr is relevant at all anymore. I will close it, we can reopen if there is still a need.

**Pull request closed without merging**